cpp/src/solver/cd_mg.cu

```diff
@@ -0,0 +1,396 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <common/cumlHandle.hpp>
+#include <common/cuml_comms_int.hpp>
+#include <common/device_buffer.hpp>
+#include <cuda_utils.cuh>
+#include <cuml/common/cuml_allocator.hpp>
+#include <cuml/linear_model/preprocess_mg.hpp>
+#include <cuml/solvers/cd_mg.hpp>
+#include <functions/softThres.cuh>
+#include <linalg/add.cuh>
+#include <linalg/eltwise.cuh>
+#include <linalg/gemm.cuh>
+#include <linalg/multiply.cuh>
+#include <linalg/subtract.cuh>
+#include <matrix/math.cuh>
+#include <matrix/matrix.cuh>
+#include <opg/linalg/mv_aTb.hpp>
```

Comment:
Future change - should we eventually change the include paths so it's clearer what comes from cumlprims_mg? Like `cumlprimsmg/opg/linalg/mv_aTb.hpp`?

Reply:
https://github.com/rapidsai/cuml/issues/2525